### PR TITLE
Torch ORTInference samples 

### PR DESCRIPTION
**Pull request by lganesan-intel**
Following samples have been added to demonstrate inference using Torch ORT Inference module and ONNXRuntime OVEP.
- NLP BERT Sequence Classification
- Vision Resnet50 Image Classification

Has a dependency on #2 to be merged prior to this as it has required bug fixes & performance/accuracy report for both the samples.

### CODE REVIEW

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    parser.add_argument('--input', type=str, default=None,
+                        help="Input sentence for prediction")
+    parser.add_argument('--input-file', type=str, default=None,
+                        help="Input file in .tsv format for prediction")
+    parser.add_argument('--provider', type=str, default="openvino",
+                        help="ONNX Runtime Execution Provider for inference")
+    parser.add_argument('--backend', type=str, 
+                        help="OpenVINO target device for Inference.")
+    parser.add_argument('--precision', type=str,
+                        help="OpenVINO target device precision for prediction")
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: Please use 'inference' across the samples.

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    print("\nPrediction took: {:.4f}s".format(total_prediction_time))
+
+def load_pred_dataset(args):
+
+    # Input can be a single sentence, list of single sentences in a .tsv file.
+    if args.input is not None:
+        sentences = [args.input]
+    elif args.input_file is not None:
+        if not os.path.exists(args.input_file):
+                raise ValueError('Invalid model path: %s' % args.input_file)
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: this should be "invalid input file path"

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                print("OpenVINOExecutionProvider is enabled with CPU and FP32 by default." +
+                    " Please specify backend and precision to override.\n")
+                args.backend = "CPU"
+                args.precision = "FP32"
+
+            if args.backend == "MYRIAD":
+                print("Execution on MYRIAD might be impacted due to large model size")
+
+    # 2. Dataloader
+    # Load input dataset/file for prediction
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: We may have to remove usage of dataset from this sample.

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                raise Exception("Invalid precision for provided backend")
+
+            if not (args.backend and args.precision):
+                print("OpenVINOExecutionProvider is enabled with CPU and FP32 by default." +
+                    " Please specify backend and precision to override.\n")
+                args.backend = "CPU"
+                args.precision = "FP32"
+
+            if args.backend == "MYRIAD":
+                print("Execution on MYRIAD might be impacted due to large model size")
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: Can add this as limitation instead of enabling this backend in sample.

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+    # 2. Dataloader
+    # Load input dataset/file for prediction
+    prediction_dataloader  = load_pred_dataset(args)
+
+    # 3. Load Model
+    # If model path is not provided, use the model
+    # Else use a pretrained model fine-tuned on CoLA dataset from huggingface model hub
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: All the comments need to be revisited since they still have references to older version.  The comment "If model path is not provided, use the model" looks incorrect.

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+
+import argparse
+import os
+import numpy as np
+import time
+import pandas as pd
+import logging
+
+from transformers import BertTokenizer, AutoConfig
+from torch.utils.data import TensorDataset, DataLoader, RandomSampler, SequentialSampler
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: May not need RandomSampler anymore. Do we still need DataLoader and TensorDataset since we are no longer using training and need data in batches?

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+from transformers import BertForSequenceClassification, BertConfig
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: May not need BertConfig since we are using AutoConfig.

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+from torch_ort import ORTInferenceModule, OpenVINOProviderOptions, DebugOptions, LogLevel
+
+ov_backend_precisions = {
+    "CPU": ["FP32"],
+    "GPU": ["FP32", "FP16"],
+    "MYRIAD": ["FP16"]
+}
+
+def get_tokenizer():
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', do_lower_case=True)
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: Can we use AutoTokenizer that's standard now.

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: Since this is new sample from Intel, we will have to update license.

File: torch_ort_inference/tests/bert_for_sequence_classification.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', do_lower_case=True)
+    return tokenizer
+
+def preprocess_input(sentences):
+    # Tokenization & Input Formatting
+    # Load the BERT tokenizer.
+    tokenizer = get_tokenizer()
+
+    # Set the max length of encoded sentence.
+    # 64 is slightly larger than the maximum training sentence length of 47...
```
Comment (saipj) [inline torch_ort_inference/tests/bert_for_sequence_classification.py]: can remove this line that refers to max. training seq length.

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+     transforms.Normalize(
+     mean=[0.485, 0.456, 0.406],
+     std=[0.229, 0.224, 0.225]
+    )])
+    return transform(input)
+
+def predict(model,image,categories):
+    # warmup
+    for _ in range(5):
+        outputs = model(image)
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py]: Can we have just 1 iteration as warm-up instead of 5?

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+ov_backend_precisions = {
+    "CPU": ["FP32"],
+    "GPU": ["FP32", "FP16"],
+    "MYRIAD": ["FP16"]
+}
+
+def download_labels(labels):
+    if not labels:
+        labelsUrl = 'https://raw.githubusercontent.com/pytorch/hub/master/imagenet_classes.txt'
+        labels = "imagenet_classes.txt"
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py:18]: The labelsUrl code may need to be moved if "imagenet_classes.txt" doesnt exist in local directory".

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+def main():
+    # 1. Basic setup
+    parser = argparse.ArgumentParser(description='PyTorch Image Classification Example')
+    parser.add_argument('--pytorch-only', action='store_true', default=False,
+                        help='disables ONNX Runtime inference')
+    parser.add_argument('--labels', type=str, default=None,
+                        help="labels file")
+    parser.add_argument('--input', type=str, required = True,
+                        help="input image for inference")
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py]: Is this path name or file name? Adding path to input image may help.

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    for i in range(top5_prob.size(0)):
+        print(categories[top5_catid[i]],top5_prob[i].item())
+
+def main():
+    # 1. Basic setup
+    parser = argparse.ArgumentParser(description='PyTorch Image Classification Example')
+    parser.add_argument('--pytorch-only', action='store_true', default=False,
+                        help='disables ONNX Runtime inference')
+    parser.add_argument('--labels', type=str, default=None,
+                        help="labels file")
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py]: Path name or just file name is enough? would be good to add more details.

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    parser.add_argument('--input', type=str, required = True,
+                        help="input image for inference")
+    parser.add_argument('--provider', type=str, default="openvino",
+                        help="ONNX Runtime Execution Provider for inference")
+    parser.add_argument('--backend', type=str,
+                        help="Backend for inference")
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py]: Publishing valid backends may help.

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    parser.add_argument('--precision', type=str,
+                        help="Precision for prediction")
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py]: Publishing valid precision values will be helpful.

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    # Convert model for evaluation
+    model.eval()
+
+    # 3. Download ImageNet labels
+    categories = download_labels(args.labels)
+    
+    # 4. Read input image and preprocess
+    if not args.input:
+        raise ValueError("Input image not provided!")
+    img = Image.open(args.input)
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py]: File check for input file path missing.

File: torch_ort_inference/tests/resnet_image_classification.py [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+def main():
+    # 1. Basic setup
+    parser = argparse.ArgumentParser(description='PyTorch Image Classification Example')
+    parser.add_argument('--pytorch-only', action='store_true', default=False,
+                        help='disables ONNX Runtime inference')
+    parser.add_argument('--labels', type=str, default=None,
+                        help="labels file")
+    parser.add_argument('--input', type=str, required = True,
+                        help="input image for inference")
+    parser.add_argument('--provider', type=str, default="openvino",
```
Comment (saipj) [inline torch_ort_inference/tests/resnet_image_classification.py]: Need not have default value. It is handled internally.

**Review by saipj — APPROVED** [review]
Looks good to me.
